### PR TITLE
Use xdg config home convention

### PR DESCRIPTION
**Pull request by sunapi386**
Implementation for https://github.com/tekezo/Karabiner-Elements/issues/112 where the config files live under `~/.config/karabiner-elements` instead of `~/.karabiner.d/configuration`.

### CODE REVIEW

File: src/share/constants.hpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ class constants final {
     return "/Library/Application Support/org.pqrs/tmp/karabiner_event_dispatcher_receiver";
   }
 
-  static const char* get_home_dot_karabiner_directory(void) {
+  static const char* get_configuration_directory(void) {
```
Comment (lgarron) [inline src/share/constants.hpp]: Note that this should first check (and use, if set) `$XDG_CONFIG_HOME` for full compatibility. (I don't personally care for that, but some users might.)

(See https://github.com/tekezo/Karabiner-Elements/issues/112#issuecomment-249102921)

Comment (sunapi386) [inline src/share/constants.hpp]: Good catch, I made the change.

**Comment by tekezo** [comment]
It's a good idea to change the configuration directory to ~/.config/karabiner!
However, this change makes a confusion since everyone change the karabiner.json manually.

I'll merge this change after Preferences UI is provided and it allows that the most user does not touch karabiner.json directly.

File: src/share/constants.hpp (hunk 1 above)
Comment (lgarron) [inline src/share/constants.hpp]: [This comment](https://github.com/tekezo/Karabiner-Elements/issues/112#issuecomment-256811473) made me realize that Karabiner-Elements as an application probably doesn't have access to shell variables.

There seems to be a way to set global shell variables, but it's not as convenient as a normal dotfile for commandline tools:
http://stackoverflow.com/a/588442

A simple compromise is to copy https://github.com/Hammerspoon/hammerspoon/pull/582 and default to `~/.config/karabiner-elements` unless a certain value is set in the Karabiner-Elements `.plist`. That doesn't scale to multiple applications the same way as `$XDG_CONFIG_HOME`, though. :-(

**Comment by lgarron** [comment]
> I'll merge this change after Preferences UI is provided and it allows that the most user does not touch karabiner.json directly.

Seems the Preferences UI is in place, and the basic functionality works!
Is now a good time to move the default config folder yet? :-P

**Comment by rafi** [comment]
please support xdg standard

**Comment by tekezo** [comment]
karabiner.json is located in ~/.config/karabiner since v0.90.70.

**Pull request closed without merging**